api/v1/verticadb_webhook.go

```diff
@@ -1341,6 +1457,85 @@ func (v *VerticaDB) checkImmutableSubclusterDuringUpgrade(oldObj *VerticaDB, all
 	return allErrs
 }
 
+// findPersistScsInSandbox returns the sandboxed subclusters that exist in both old and new vdb
+func (v *VerticaDB) findPersistScsInSandbox(oldObj *VerticaDB) []string {
+	oldSandboxes := oldObj.Spec.Sandboxes
+	newSandboxes := v.Spec.Sandboxes
+	sandboxScMap := make(map[string][]SubclusterName)
+	for _, sandbox := range oldSandboxes {
+		sandboxScMap[sandbox.Name] = sandbox.Subclusters
+	}
+	persistScs := []string{}
+	for _, sandbox := range newSandboxes {
+		if oldScs, ok := sandboxScMap[sandbox.Name]; ok {
+			oldScMap := make(map[string]any)
+			for _, oldSc := range oldScs {
+				oldScMap[oldSc.Name] = struct{}{}
+			}
+			for _, newSc := range sandbox.Subclusters {
+				if _, ok := oldScMap[newSc.Name]; ok {
+					persistScs = append(persistScs, newSc.Name)
+				}
+			}
+		}
+	}
+	return persistScs
+}
+
+// checkImmutableSubclusterInSandbox ensures we do not scale and remove any subcluster that is in a sandbox
+func (v *VerticaDB) checkImmutableSubclusterInSandbox(oldObj *VerticaDB, allErrs field.ErrorList) field.ErrorList {
+	// if either old vdb or new vdb does not have any sandboxes, skip this check
+	if len(oldObj.Spec.Sandboxes) == 0 || len(v.Spec.Sandboxes) == 0 {
+		return allErrs
+	}
+
+	persistScs := v.findPersistScsInSandbox(oldObj)
+	oldScMap := oldObj.GenSubclusterMap()
+	newScMap := v.GenSubclusterMap()
+	path := field.NewPath("spec").Child("subclusters")
+
+	// check if the sizes of subclusters in a sandbox get changed
+	for _, sc := range persistScs {
+		oldSc, hasOldSc := oldScMap[sc]
+		newSc, hasNewSc := newScMap[sc]
+		if !hasOldSc || !hasNewSc {
+			err := field.Invalid(path,
+				v.Spec.Subclusters,
+				fmt.Sprintf("Subcluster %s in vdb.spec.sandboxes cannot be found in vdb.spec.subclusters", sc))
+			allErrs = append(allErrs, err)
+			continue
+		}
+		if oldSc.Size != newSc.Size {
+			err := field.Invalid(path,
+				v.Spec.Subclusters,
+				fmt.Sprintf("Cannot change the size of subcluster %q when it is in a sandbox", sc))
+			allErrs = append(allErrs, err)
+			continue
+		}
+	}
+
+	// find subclusters that are sandboxed in old vdb but removed in new vdb
+	oldSandboxes := oldObj.Spec.Sandboxes
+	scInSandbox := make(map[string]any)
+	for _, sandbox := range oldSandboxes {
+		for _, sc := range sandbox.Subclusters {
+			scInSandbox[sc.Name] = struct{}{}
+		}
+	}
+	removedScs := vutil.MapKeyDiff(oldScMap, newScMap)
+	for _, sc := range removedScs {
+		if _, ok := scInSandbox[sc]; ok {
+			err := field.Invalid(path,
+				v.Spec.Subclusters,
```

Comment:
general: this may generate a lot of output in the error because the message will print out the entire subcluster list. We may want to only include the offending subcluster. Something that we find in one of the subcluster maps should be fine.